drivers: pwm: sam: update to support Microchip sama7g5

TonyHan11 opened the pull request:
The following changes are addressed in this PR:
- update the pwm driver for supporting Microchip sama7g5
- enable `samples/basic/blinky_pwm` for sama7g54_ek board

TonyHan11 committed:
soc: microchip: sam: update MMU for sama7g5 PWM

When the PWM is activated in the DT, configure it's register region
with strong ordered, read and write access.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: pwm: pwm_sam: update to support Soc SAMA7G5

Redefine some macros due to different naming scheme in the header file.
Get the clock rate from the device tree when 'SOC_ATMEL_SAM_MCK_FREQ_HZ'
is not defined.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
dtsi: microchip: sam: add PWM nodes for sama7g5

Add the PWM node to sama7g5.dtsi file.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
samples: pwm: add sama7g54_ek.overlay file to the blink_pwm sample

The sama7g54-ek has the LEDs those can be driven by GPIO or PWM.

Signed-off-by: Tony Han <tony.han@microchip.com>

nandojve commented on drivers/pwm/pwm_sam.c:
```diff
@@ -17,6 +17,14 @@
 
 LOG_MODULE_REGISTER(pwm_sam, CONFIG_PWM_LOG_LEVEL);
 
+/* SAMA7G5 uses a slightly different naming scheme in header file component/pmc.h */
+#ifdef _SAMA7G5_PWM_COMPONENT_H_
+#undef PWM_CMR_CPOL
+#define PWM_CMR_CPOL    PWM_CMR_CPOL_Msk
+#define PWMCHNUM_NUMBER PWM_CH_NUM_NUMBER
+typedef pwm_registers_t Pwm;
+#endif
```
So, we start with fixups again.

CC: @NhMchp

nandojve commented on boards/microchip/sam/sama7g54_ek/sama7g54_ek.dts:
```diff
@@ -96,8 +110,27 @@
 };
+
+&pwm {
+	status = "disabled"; /* Conflict with leds. */
+
```
It must be disabled by default inside soc.dtsi.
```diff
-	status = "disabled"; /* Conflict with leds. */
-
```

nandojve commented on soc/microchip/sam/sama7g5/soc.c:
```diff
@@ -33,6 +33,10 @@ static const struct arm_mmu_region mmu_regions[] = {
 			      MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),
 
+	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(atmel_sam_pwm),
+		   (MMU_REGION_FLAT_ENTRY("pwm", PWM_BASE_ADDRESS, 0x500,
+					 MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),))
```
```diff
		   (MMU_REGION_FLAT_ENTRY("pwm", PWM_BASE_ADDRESS, 0x500,
-					 MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),))
+					  MT_STRONGLY_ORDERED | MPERM_R | MPERM_W),))
```

nandojve commented on drivers/pwm/pwm_sam.c:
```diff
@@ -17,6 +17,14 @@
 
 LOG_MODULE_REGISTER(pwm_sam, CONFIG_PWM_LOG_LEVEL);
 
+/* SAMA7G5 uses a slightly different naming scheme in header file component/pmc.h */
```
The driver commit is after dts commit, not before.

nandojve commented on samples/basic/blinky_pwm/boards/sama7g54_ek.overlay:
```diff
@@ -0,0 +1,21 @@
+ */
+
+/ {
+	aliases {
+		pwm-led0 = &pwm_led_green;
+	};
```
This can be inside board

TonyHan11 committed:
boards: microchip: sam: add PWM to sama7g54-ek dts and yaml files

Add PWM and pwmled nodes to sama7g54_ek.dts file.
Add pwm to sama7g54_ek.yaml support list.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented:
@nandojve , thank you for your comments and suggestions.
Updated with the following:
- put the dtsi commit before the commit for the driver
- board dts in a separate commit after the driver
- remove `status = "disabled"` for pwm and pwmleds nodes
- align the code for enable MMU for PWM
- move `aliases` from overlay to board dts

TonyHan11 commented:
> @nandojve , thank you for your comments and suggestions. Updated with the following:
> 
>     * put the dtsi commit before the commit for the driver
> 
>     * board dts in a separate commit after the driver
> 
>     * remove `status = "disabled"` for pwm and pwmleds nodes
> 
>     * align the code for enable MMU for PWM
> 
>     * move `aliases` from overlay to board dts

revert `status = "disabled"` for node pwmleds as pwm is default disabled.

dsseng commented on drivers/pwm/pwm_sam.c:
```diff
@@ -44,8 +52,21 @@ static int sam_pwm_get_cycles_per_sec(const struct device *dev,
 
-	*cycles = SOC_ATMEL_SAM_MCK_FREQ_HZ /
-		  ((1 << prescaler) * divider);
+#ifdef SOC_ATMEL_SAM_MCK_FREQ_HZ
+	uint32_t rate = SOC_ATMEL_SAM_MCK_FREQ_HZ;
+#else
+	uint32_t rate;
+	int ret;
+
+	ret = clock_control_get_rate(SAM_DT_PMC_CONTROLLER,
+				     (clock_control_subsys_t)&config->clock_cfg,
+				     &rate);
+	if (ret < 0) {
+		return ret;
+	}
+#endif
```
Couldn't this all be handled by e.g. a clock reference in the PWM DT node, referring to the right clock source in each SoC's DT?

TonyHan11 replied:
`SOC_ATMEL_SAM_MCK_FREQ_HZ` is the frequency for MCUs and it is defined as `SOC_ATMEL_SAM_HCLK_FREQ_HZ`:
```
soc/atmel/sam/sam3x/soc.h:48:#define SOC_ATMEL_SAM_MCK_FREQ_HZ SOC_ATMEL_SAM_HCLK_FREQ_HZ
soc/atmel/sam/sam4e/soc.h:47:#define SOC_ATMEL_SAM_MCK_FREQ_HZ SOC_ATMEL_SAM_HCLK_FREQ_HZ
soc/atmel/sam/sam4l/soc.h:67:#define SOC_ATMEL_SAM_MCK_FREQ_HZ       SOC_ATMEL_SAM_HCLK_FREQ_HZ
soc/atmel/sam/sam4s/soc.h:61:#define SOC_ATMEL_SAM_MCK_FREQ_HZ SOC_ATMEL_SAM_HCLK_FREQ_HZ
soc/atmel/sam/samx7x/soc.h:110:#define SOC_ATMEL_SAM_MCK_FREQ_HZ \
soc/atmel/sam/samx7x/soc.h-111- (SOC_ATMEL_SAM_HCLK_FREQ_HZ / CONFIG_SOC_ATMEL_SAM_MDIV)
```
`SOC_ATMEL_SAM_HCLK_FREQ_HZ` is defined as `ATMEL_SAM_DT_CPU_CLK_FREQ_HZ`:
```
soc/atmel/sam/sam3x/soc.h:45:#define SOC_ATMEL_SAM_HCLK_FREQ_HZ ATMEL_SAM_DT_CPU_CLK_FREQ_HZ
soc/atmel/sam/sam4e/soc.h:44:#define SOC_ATMEL_SAM_HCLK_FREQ_HZ ATMEL_SAM_DT_CPU_CLK_FREQ_HZ
soc/atmel/sam/sam4l/soc.h:64:#define SOC_ATMEL_SAM_HCLK_FREQ_HZ      ATMEL_SAM_DT_CPU_CLK_FREQ_HZ
soc/atmel/sam/sam4s/soc.h:58:#define SOC_ATMEL_SAM_HCLK_FREQ_HZ ATMEL_SAM_DT_CPU_CLK_FREQ_HZ
soc/atmel/sam/samx7x/soc.h:107:#define SOC_ATMEL_SAM_HCLK_FREQ_HZ ATMEL_SAM_DT_CPU_CLK_FREQ_HZ
```

Finally `ATMEL_SAM_DT_CPU_CLK_FREQ_HZ` comes from [DT (clock-frequency of cpu0)](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/atmel/sam/common/atmel_sam_dt.h#L18-L19).

It's better to handle it in another PR, thanks.

TonyHan11 commented:
Rebased on the latest `main` branch for solving conflicts in device tree.

TonyHan11 commented on boards/microchip/sam/sama7g54_ek/sama7g54_ek.dts:
```diff
@@ -21,6 +22,7 @@
 	aliases {
 		led0 = &led_green;
 		sw0 = &button_user;
+		pwm-led0 = &pwm_led_green;
```
Updated the commit message and added pwm to the support list in sama7g54_ek.yaml file, thank you very much.

TonyHan11 commented:
The commits have conflicts due to the sdmmc is merged.
Rebased with following conflicts solved:
- the nodes in `dts/arm/microchip/sam/sama7g5.dtsi`
- MMU configuration in `soc/microchip/sam/sama7g5/soc.c`
- the nodes in `boards/microchip/sam/sama7g54_ek/sama7g54_ek.dts`
- the support list in `boards/microchip/sam/sama7g54_ek/sama7g54_ek.yaml`

TonyHan11 commented:
Hi @pdgendt , @nandojve ,
I've rebased the branch with conflict solved. Could you kindly review and approve it again? Thanks a lot.
